### PR TITLE
Skip materialized view checking if source table does not exist

### PR DESCRIPTION
**Pull request by MikhailBurdukov**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed `UNKNOWN_TABLE` exception when attaching to a materialized view that has dependent tables that are not available.
This might  be useful when trying to restore state from a backup.

Steps to reproduce:
```
CREATE DATABASE test_db;

CREATE TABLE test_db.table (n Int32, s String) ENGINE MergeTree PARTITION BY n ORDER BY n;

CREATE TABLE test_db.mview_backend (n Int32, n2 Int64) ENGINE MergeTree PARTITION BY n ORDER BY n;

CREATE MATERIALIZED VIEW test_db.mview TO test_db.mview_backend AS SELECT n, n * n AS "n2" FROM test_db.table;

DROP TABLE test_db.table;

DETACH TABLE test_db.mview;

ATTACH TABLE test_db.mview;
```
```
Received exception from server (version 22.11.5):
Code: 60. DB::Exception: Received from localhost:9001. DB::Exception: Table test_db.table doesn't exist. (UNKNOWN_TABLE)
```
Expected behavior: attaching materialized view succeeds regardless of whether it's valid and references existing table, or not. It's important for backup-restore scenarios.

With ordinary view  works as expected.
```
CREATE DATABASE test_db;

CREATE TABLE test_db.table (n Int32, s String) ENGINE MergeTree PARTITION BY n ORDER BY n;

CREATE VIEW test_db.view AS SELECT n, n * n AS "n2" FROM test_db.table;

DROP TABLE test_db.table;

DETACH TABLE test_db.view;

ATTACH TABLE test_db.view;
```
```
Ok.
```

### CODE REVIEW

File: src/Interpreters/InterpreterCreateQuery.h [hunk 1]
```diff
@@ -95,6 +95,8 @@ class InterpreterCreateQuery : public IInterpreter, WithMutableContext
     BlockIO doCreateOrReplaceTable(ASTCreateQuery & create, const InterpreterCreateQuery::TableProperties & properties);
     /// Inserts data in created table if it's CREATE ... SELECT
     BlockIO fillTableIfNeeded(const ASTCreateQuery & create);
+    /// Check type compatible for materialized dest table and select columns
+    void checkTypecompatibleForMaterializeView(const ASTCreateQuery & create);
```
Comment (kssenii) [inline src/Interpreters/InterpreterCreateQuery.h:99]:
```suggestion
    void checkTypeCompatibleForMaterializeView(const ASTCreateQuery & create);
```

File: src/Core/Settings.h [hunk 2]
```diff
@@ -467,6 +467,7 @@ class IColumn;
     M(Int64, max_partitions_to_read, -1, "Limit the max number of partitions that can be accessed in one query. <= 0 means unlimited.", 0) \
     M(Bool, check_query_single_value_result, true, "Return check query result as single 1/0 value", 0) \
     M(Bool, allow_drop_detached, false, "Allow ALTER TABLE ... DROP DETACHED PART[ITION] ... queries", 0) \
+    M(Bool, skip_materialized_view_checking_if_source_table_not_exist, false, "Allow attaching to a materialized view even if dependent tables are inaccessible.", 0) \
```
[the remark this answers is not in the document]
Comment (kssenii) [reply src/Core/Settings.h:470]: Agree that setting is probably redundant, ok to remove. (Though if not removing it - in my opinion its name should end with `_on_attach` )

[the remark this answers is not in the document]
Comment (MikhailBurdukov) [reply src/Core/Settings.h]: > We should always skip the check on ATTACH query, the check makes sense for CREATE only

If I understood you correctly, we should check type compatibility only in cases of `CREATE`. I have reverted changes and added a simple check that the query is not attached.

**Review by kssenii — APPROVED** [review]

**Comment by kssenii** [comment]
> AST fuzzer (ubsan) — Received signal 11

https://github.com/ClickHouse/ClickHouse/issues/48094

**Pull request merged**